### PR TITLE
Configurable wait for completion timeout

### PR DESCRIPTION
**Pull request by benediktibk**
Depending on the ADC implementation it might happen that the driver is waiting on an external interrupt. If this interrupt gets lost, for instance due to a race condition with an external port expander, the system will get stuck.
Making this configurable allows the user to recover from such an error.

### CODE REVIEW

**Comment by decsny** [comment]
since this is describing a particular type of hardware configuration, should it be driven by devicetree in some way?

**Comment by benediktibk** [comment]
Having this timeout in the devicetree would allow it to be configured differently for each ADC driver and instance. But it feels kinda weird for me to define a software timeout in the hardware description. Are there similar examples around in zephyr?

**Comment by decsny** [comment]
> Having this timeout in the devicetree would allow it to be configured differently for each ADC driver and instance. But it feels kinda weird for me to define a software timeout in the hardware description. Are there similar examples around in zephyr?

What I was referring to as far as what to put in DT was not the software timeout itself but the description of the hardware setup where the ADC waits on an external interrupt. Because if this is how the hardware works it seems like it might be better to describe it in DT instead of setting a kconfig to -1

**Comment by benediktibk** [comment]
> > Having this timeout in the devicetree would allow it to be configured differently for each ADC driver and instance. But it feels kinda weird for me to define a software timeout in the hardware description. Are there similar examples around in zephyr?
> 
> What I was referring to as far as what to put in DT was not the software timeout itself but the description of the hardware setup where the ADC waits on an external interrupt. Because if this is how the hardware works it seems like it might be better to describe it in DT instead of setting a kconfig to -1

I see. Maybe something in the spirit of ADC_CONTEXT_USES_KERNEL_TIMER? Because in the devicetree there is already a definition if the driver will wait on a hardware interrupt, namely the existence or non-existence of a data ready GPIO. Which is of course driver specific.

    #define ADC_CONTEXT_USES_KERNEL_TIMER 1
    #define ADC_CONTEXT_WAIT_ON_HARDWARE_INTERRUPT 1
    #include "adc_context.h"

    config ADC_WAIT_FOR_COMPLETION_TIMEOUT_IN_MS
        int "timeout for wait for completion of an ADC operation in ms"
        default 10
        help
        An ADC driver will wait this time in ms until an operation
        is completed. This is only used if the drivers sets 
        ADC_CONTEXT_WAIT_ON_HARDWARE_INTERRUPT.

But this is again kind of problematic, because there might be multiple instances of the same driver, where one have a GPIO and the others do not have. As the wait for completion is happening within adc_context.h I can do this at the moment either for all instances of a driver, or for none. The only way around this, which I currently see, is extending the adc_context with a field, for instance containing the timeout. Which then could be set in the init function specifically for each driver instance.

File: drivers/adc/adc_context.h [hunk 1]
```diff
@@ -48,6 +48,12 @@ static void adc_context_disable_timer(struct adc_context *ctx);
 static void adc_context_on_complete(struct adc_context *ctx, int status);
 #endif /* ADC_CONTEXT_ENABLE_ON_COMPLETE */
 
+#if CONFIG_ADC_WAIT_FOR_COMPLETION_TIMEOUT_IN_MS == -1
+#define ADC_CONTEXT_WAIT_FOR_COMPLETION_TIMEOUT K_FOREVER
+#else
+#define ADC_CONTEXT_WAIT_FOR_COMPLETION_TIMEOUT K_MSEC(CONFIG_ADC_WAIT_FOR_COMPLETION_TIMEOUT_IN_MS)
+#endif
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/adc_context.h:55]: Sounds reasonable to me

[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/adc_context.h]: done

File: drivers/adc/adc_lmp90xxx.c [hunk 2]
```diff
@@ -595,7 +595,7 @@ static int lmp90xxx_adc_read_channel(const struct device *dev,
 	}
 
 	if (LMP90XXX_HAS_DRDYB(config)) {
-		k_sem_take(&data->drdyb_sem, K_FOREVER);
+		k_sem_take(&data->drdyb_sem, ADC_CONTEXT_WAIT_FOR_COMPLETION_TIMEOUT);
```
Comment (anangl) [inline drivers/adc/adc_lmp90xxx.c:598]: It is still the default `K_FOREVER` here. What was the intention of that change?
For `ADC_CONTEXT_WAIT_FOR_COMPLETION_TIMEOUT` defined to something else, the return value should not be just ignored.

Comment (benediktibk) [reply drivers/adc/adc_lmp90xxx.c]: done

**Review by anangl — DISMISSED** [review]

**Review by aasinclair — DISMISSED** [review]

File: drivers/adc/Kconfig.ads114s0x [hunk 3]
```diff
@@ -34,3 +34,10 @@ config ADC_ADS114S0X_GPIO
 
 	  The GPIO functionality is handled by the ADS114S0x GPIO
 	  driver.
+
+config ADC_ADS114S0X_WAIT_FOR_COMPLETION_TIMEOUT_IN_MS
```
Comment (fabiobaltieri) [inline drivers/adc/Kconfig.ads114s0x:38]: nit: this should probably be `config ADC_ADS114S0X_WAIT_FOR_COMPLETION_TIMEOUT_MS`, that's how all other time Kconfig entries are defined.

File: drivers/adc/Kconfig.ads114s0x [hunk 4]
```diff
@@ -34,3 +34,10 @@ config ADC_ADS114S0X_GPIO
...
+	int "timeout for wait for completion of a read in ms"
```
Comment (fabiobaltieri) [inline drivers/adc/Kconfig.ads114s0x:39]: s/timeout/Timeout/

File: drivers/adc/Kconfig.ads114s0x (hunk 3 above)
Comment (benediktibk) [reply drivers/adc/Kconfig.ads114s0x]: done

**Review by fabiobaltieri — APPROVED** [review]

File: drivers/adc/Kconfig.ads114s0x (hunk 4 above)
Comment (benediktibk) [reply drivers/adc/Kconfig.ads114s0x]: done

**Review by anangl — APPROVED** [review]

**Review by aasinclair — APPROVED** [review]

**Pull request merged**